examples/coap-server.c

```diff
@@ -1835,40 +1835,53 @@ static uint8_t *read_file_mem(const char* file, size_t *length) {
   return buf;
 }
 
+static void
+update_pki_key(coap_dtls_key_t *dtls_key, const char *key_name,
+               const char *cert_name, const char *ca_name) {;
+  memset (dtls_key, 0, sizeof(*dtls_key));
+  if ((key_file && strncasecmp (key_name, "pkcs11:", 7) == 0) ||
```

Comment:
Checking `key_file`, `cert_file`, and `ca_file` in the previous version has protected the `strncasecmp()`. Is this change of behavior intentional?

Moreover, I have difficulties to see if the changed if-else-structure still handles the `use_pem_buf` cases correctly.

Reply:
Previously, `verify_pki_sni_callback()` and `setup_pki()` both set up the `dtls_pki.pki_key` structure.  However, the `verify_pki_sni_callback()` variant did not handle the situation when `is_rpk_not_cert `was set.  I moved the `setup_pki()` variant into a new function` update_pki_key()` unchanged other than `dtls_pki.pki_key.` had become `dtls_key->` and `ca_mem`, `cert_mem `and `key_mem` were always freed and built as appropriate.  I then called this new function from the appropriate places.

I can't see where I have changed the protection use of` strncasecmp()`.

Reply:
Previously, we had the pattern `if (X && strncasecmp(X,...) ...`, now it is `if (X && strncasecmp(Y,...) ...`
Maybe this is a result from inconsistent renaming from *_file to *_name?

Reply:
OK, now understand.  Code change just pushed to fix this.

Reply:
I still have troubles understanding the new structure. If `-m` is passed on the command line weird things will happen if one of `key_file`, `ca_file` or `cert_file` start with the prefix "`pkcs11:`" I think that `-m` should override automatic URI schema detection as was done previously.

Reply:
As per HOWTO.pkcs11 at the end
````
#
# Client and Server using RPK (GnuTLS only)
#
coap-server -M 'pkcs11:token=token-0;object=server-key' -J 1234 -v9 -n
# and
coap-client -M 'pkcs11:token=token-0;object=client-key' -J 1234 -v9 coaps://[::1]
````
which works as I expect it to.  True for code prior to this PR as well as this PR.  Note that it is upper` -M` option.

Reply:
The question is what happens with `-m`?

```
coap-client -m -M 'pkcs11:token=token-0;object=client-key' -J 1234 -v9 coaps://[::1]
```

Reply:
````
        -m method       Request method (get|put|post|delete|fetch|patch|ipatch),
                        default is 'get'
````
I am surprised it does not complain about `-M` not being get/post etc., but understand it complaining that `'pkcs11:token=token-0;object=client-key'` is an invalid CoAP URI.

Reply:
I am sorry, this was a copy&paste error. The issue I see is with coap-server, not coap-client. In coap-server, `-m` is used to set `use_pem_buf`. (Unfortunately, I cannot test this by myself because the recipe described in HOWTO.pkcs11 does not work on my system, and I have not yet been able to find out why.)

```
coap-server -m -M 'pkcs11:token=token-0;object=server-key' -J 1234 -v9 -n
```

Reply:
OK - I also need to carefully read the context (about coap-server) and the command (coap-client) conflicts!

From the GnuTLS perspective, if `pkcs11:` URL prefix, it does a` gnutls_certificate_set_rawpk_key_file()` for RPK as `-M` is used or `gnutls_certificate_set_x509_key_file() `for PKI if `-c `is used (which are both 'file' based).  See line 1314 src/coap_gnutls.c.  So the `-m` is superfluous in this case and there is no need to map the file into memory.

This works for me in my environment with softhsm2 set up as per document ion for setting up in PR #670.

Reply:
Okay, if `-m` is superfluous here, this new structure would work. The Howto.pkcs11 documentation in PR #670 indeed worked for me as well, so I can start testing.